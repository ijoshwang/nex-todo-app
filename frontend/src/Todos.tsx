import React, { useState } from 'react'
import { DeleteOutlined, EditOutlined } from '@ant-design/icons'
import { Button, Input, List, Space, Typography } from 'antd'

interface Task {
  id: number
  name: string
}

const Component: React.FC = () => {
  const [tasks, setTasks] = useState<Task[]>([
    { id: 1, name: 'Finish project proposal' },
    { id: 2, name: 'Schedule team meeting' },
    { id: 3, name: 'Buy groceries' },
  ])
  const [newTask, setNewTask] = useState<string>('')
  const [editingTask, setEditingTask] = useState<number | null>(null)
  const [editedTask, setEditedTask] = useState<string>('')

  const handleAddTask = () => {
    if (newTask.trim() !== '') {
      const newId =
        tasks.length > 0 ? Math.max(...tasks.map((t) => t.id)) + 1 : 1
      setTasks([...tasks, { id: newId, name: newTask }])
      setNewTask('')
    }
  }

  const handleEditTask = (task: Task) => {
    setEditingTask(task.id)
    setEditedTask(task.name)
  }

  const handleUpdateTask = () => {
    if (editedTask.trim() !== '') {
      setTasks(
        tasks.map((task) =>
          task.id === editingTask ? { ...task, name: editedTask } : task
        )
      )
      setEditingTask(null)
      setEditedTask('')
    }
  }

  const handleDeleteTask = (id: number) => {
    setTasks(tasks.filter((task) => task.id !== id))
  }

  return (
    <div
      style={{
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'center',
        height: '100vh',
      }}
    >
      <div
        style={{
          width: '100%',
          maxWidth: '400px',
          padding: '24px',
          backgroundColor: '#ffffff',
          borderRadius: '8px',
          boxShadow: '0 2px 8px rgba(0, 0, 0, 0.15)',
        }}
      >
        <Typography.Title level={2} style={{ textAlign: 'center' }}>
          To-Do List
        </Typography.Title>
        <Space style={{ marginBottom: '16px', width: '100%' }}>
          <Input
            placeholder="Add a new task"
            value={newTask}
            onChange={(e) => setNewTask(e.target.value)}
            style={{ flex: 1 }}
          />
          <Button type="primary" onClick={handleAddTask}>
            Add
          </Button>
        </Space>
        <List
          dataSource={tasks}
          renderItem={(task) => (
            <List.Item
              actions={[
                editingTask === task.id ? (
                  <Button type="link" onClick={handleUpdateTask}>
                    Update
                  </Button>
                ) : (
                  <Button
                    type="link"
                    icon={<EditOutlined />}
                    onClick={() => handleEditTask(task)}
                  />
                ),
                <Button
                  type="link"
                  icon={<DeleteOutlined />}
                  onClick={() => handleDeleteTask(task.id)}
                />,
              ]}
            >
              {editingTask === task.id ? (
                <Input
                  value={editedTask}
                  onChange={(e) => setEditedTask(e.target.value)}
                  style={{ flex: 1 }}
                />
              ) : (
                <Typography.Text>{task.name}</Typography.Text>
              )}
            </List.Item>
          )}
        />
      </div>
    </div>
  )
}

export default Component
