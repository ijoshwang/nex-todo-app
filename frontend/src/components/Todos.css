#todos {
  background-color: #fff;
  border-radius: 10px;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.15);
  padding: 24px;
  min-height: 200px;
  max-width: 720px;
  margin: 0 auto;
}

#todos h3 {
  font-size: 20px;
  margin: 0 0 20px;
  padding: 0;
  text-align: center;
}

#todos .list {
  font-size: 18px;
  line-height: 2;
  margin: 0;
  padding: 0;
  list-style: none;
}

#todos .list li {
  border-bottom: 1px solid #eee;
  padding: 10px 0;
  cursor: pointer;
}

#todos .list li:hover {
  text-shadow: 1px 1px 2px rgba(0, 0, 0, 0.2);
}

#todos .list li span span:hover {
  text-shadow: 1px 2px 2px rgba(0, 0, 0, 0.4);
}

#todos .list li.is-completed {
  opacity: 0.7;
  text-decoration: line-through;
}

#todos .item {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

#todos .add-item {
  cursor: pointer;
  display: flex;
  align-items: center;
  font-size: 18px;
  line-height: 2;
  padding: 10px 0;
}

#todos .edit-item {
  display: flex;
  align-items: center;
  width: 100%;
  position: relative;
}

#todos .edit-input {
  font-size: 18px;
  line-height: 2;
  margin: 0 10px 0 14px;
  padding: 0 0 0 10px;
  /* background-color: #fff;
  border-radius: 4px;
  box-shadow:
    0 4px 6px rgba(0, 0, 0, 0.1),
    0 1px 3px rgba(0, 0, 0, 0.08); */
}

#todos .edit-item .error-message {
  font-size: 14px;
  color: tomato;
  position: absolute;
  top: 50%;
  right: 80px;
  transform: translateY(-50%);
}
