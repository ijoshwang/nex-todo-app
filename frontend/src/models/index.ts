export interface DutyDTO {
  id: string
  name: string
  created_at: string
  updated_at: string
  is_completed: boolean
}

export interface Duty {
  id: string
  name: string
  createdAt: string
  updatedAt: string
  isCompleted: boolean
}
