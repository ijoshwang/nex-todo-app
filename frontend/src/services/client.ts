const API_BASE_URL =
  import.meta.env.VITE_API_BASE_URL || 'http://localhost:3000'

interface RequestOptions extends RequestInit {
  url: string
}

class HTTPError extends Error {
  status: number
  response: unknown

  constructor(message: string, status: number, response: unknown) {
    super(message)
    this.status = status
    this.response = response
  }
}

class APIClient {
  baseURL: string

  constructor(baseURL: string) {
    this.baseURL = baseURL
  }

  private mergeHeaders(headers?: HeadersInit): HeadersInit {
    const defaultHeaders = {
      'Content-Type': 'application/json',
    }

    return headers ? { ...defaultHeaders, ...headers } : defaultHeaders
  }

  async request<T>({ url, headers, ...options }: RequestOptions): Promise<T> {
    const response = await fetch(`${this.baseURL}${url}`, {
      ...options,
      headers: this.mergeHeaders(headers),
    })

    if (!response.ok) {
      const responseBody = await response.json()
      throw new HTTPError('HTTP Error', response.status, responseBody)
    }

    return response.json()
  }

  get<T>(url: string, headers?: HeadersInit): Promise<T> {
    return this.request<T>({
      url,
      method: 'GET',
      headers,
    })
  }

  post<T>(url: string, data: unknown, headers?: HeadersInit): Promise<T> {
    return this.request<T>({
      url,
      method: 'POST',
      headers,
      body: JSON.stringify(data),
    })
  }

  put<T>(url: string, data: unknown, headers?: HeadersInit): Promise<T> {
    return this.request<T>({
      url,
      method: 'PUT',
      headers,
      body: JSON.stringify(data),
    })
  }

  delete<T>(url: string, headers?: HeadersInit): Promise<T> {
    return this.request<T>({
      url,
      method: 'DELETE',
      headers,
    })
  }
}

export const apiClient = new APIClient(API_BASE_URL)
