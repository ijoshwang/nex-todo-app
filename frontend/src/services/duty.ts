import { apiClient } from './client'

export interface Duty {
  id: string
  name: string
  isCompleted: boolean
  created_at: string
  updated_at: string
}

export const getDuties = async (): Promise<Duty[]> => {
  try {
    return await apiClient.get<Duty[]>('/duties')
  } catch (error) {
    // Handle the error or rethrow it to be handled by the caller
    throw new Error('Failed to fetch duties')
  }
}

export const getDutyById = async (id: string): Promise<Duty> => {
  try {
    return await apiClient.get<Duty>(`/duties/${id}`)
  } catch (error) {
    throw new Error(`Failed to fetch duty with id ${id}`)
  }
}

export const createDuty = async (name: string): Promise<Duty> => {
  try {
    return await apiClient.post<Duty>('/duties', { name, isCompleted: false })
  } catch (error) {
    throw new Error('Failed to create duty')
  }
}

export const updateDuty = async (
  id: string,
  name: string,
  isCompleted: boolean
): Promise<Duty> => {
  try {
    return await apiClient.put<Duty>(`/duties/${id}`, { name, isCompleted })
  } catch (error) {
    throw new Error(`Failed to update duty with id ${id}`)
  }
}

export const deleteDuty = async (id: string): Promise<void> => {
  try {
    return await apiClient.delete<void>(`/duties/${id}`)
  } catch (error) {
    throw new Error(`Failed to delete duty with id ${id}`)
  }
}
