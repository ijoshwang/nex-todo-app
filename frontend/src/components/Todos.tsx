import { useEffect, useState } from 'react'
import { Spin } from 'antd'

import { Duty, getDuties } from '../services/duty'

import AddItem from './AddItem'
import Item from './Item'

import './Todos.css'

export default function Todos() {
  const [items, setItems] = useState<Duty[]>([])
  const [loading, setLoading] = useState<boolean>(true)
  const [error, setError] = useState<string | null>(null)

  const fetchItems = async () => {
    try {
      const duties = await getDuties()
      setItems(duties)
      setLoading(false)
    } catch (err) {
      setError('Failed to load duties')
      setLoading(false)
    }
  }

  useEffect(() => {
    fetchItems()
  }, [])

  if (loading) {
    return <Spin />
  }

  if (error) {
    return <div>{error}</div>
  }

  return (
    <div id="todos">
      <h3>NEX TODO</h3>
      <ul className="list">
        {items.map((item) => (
          <Item key={item.id} item={item} saveCallback={fetchItems} />
        ))}
      </ul>
      <AddItem saveCallback={fetchItems} />
    </div>
  )
}
