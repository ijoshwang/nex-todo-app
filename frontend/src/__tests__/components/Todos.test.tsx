import { act } from 'react'
import { render, screen, waitFor } from '@testing-library/react'

import Todos from '@/components/Todos'
import { Duty } from '@/models'
import { getDuties } from '@/services/duty'

// Mock the getDuties service
jest.mock('@/services/duty', () => ({
  getDuties: jest.fn(),
}))

const mockDuties: Duty[] = [
  {
    id: '1',
    name: 'Duty 1',
    isCompleted: false,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
  },
  {
    id: '2',
    name: 'Duty 2',
    isCompleted: true,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
  },
]

describe('Todos Component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('renders Todos component', async () => {
    ;(getDuties as jest.Mock).mockResolvedValue(mockDuties)
    let container: HTMLElement | null = null
    await act(async () => {
      const { container: renderedContainer } = render(<Todos />)
      container = renderedContainer
    })
    expect(screen.getByText('NEX TODO')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  test('displays loading spinner while fetching duties', async () => {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    ;(getDuties as jest.Mock).mockReturnValue(new Promise(() => {}))
    let container: HTMLElement | null = null
    await act(async () => {
      const { container: renderedContainer } = render(<Todos />)
      container = renderedContainer
    })
    expect(screen.getByRole('alert')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  test('displays error message if fetching duties fails', async () => {
    ;(getDuties as jest.Mock).mockRejectedValue(
      new Error('Failed to load duties')
    )
    let container: HTMLElement | null = null
    await act(async () => {
      const { container: renderedContainer } = render(<Todos />)
      container = renderedContainer
    })
    await waitFor(() => {
      expect(screen.getByText('Failed to load duties')).toBeInTheDocument()
    })
    expect(container).toMatchSnapshot()
  })

  test('displays duties after fetching', async () => {
    ;(getDuties as jest.Mock).mockResolvedValue(mockDuties)
    let container: HTMLElement | null = null
    await act(async () => {
      const { container: renderedContainer } = render(<Todos />)
      container = renderedContainer
    })
    await waitFor(() => {
      expect(screen.getByText('Duty 1')).toBeInTheDocument()
      expect(screen.getByText('Duty 2')).toBeInTheDocument()
    })
    expect(container).toMatchSnapshot()
  })

  test('renders AddItem component', async () => {
    ;(getDuties as jest.Mock).mockResolvedValue(mockDuties)
    let container: HTMLElement | null = null
    await act(async () => {
      const { container: renderedContainer } = render(<Todos />)
      container = renderedContainer
    })
    await waitFor(() => {
      expect(screen.getByText('Add a new item')).toBeInTheDocument()
    })
    expect(container).toMatchSnapshot()
  })
})
