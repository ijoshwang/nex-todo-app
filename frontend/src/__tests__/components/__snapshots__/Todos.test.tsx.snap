// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos Component displays duties after fetching 1`] = `
<div>
  <div
    id="todos"
  >
    <h3>
      NEX TODO
    </h3>
    <ul
      class="list"
    >
      <li
        class=""
      >
        <div
          class="item"
        >
          <label
            class="ant-checkbox-wrapper css-dev-only-do-not-override-1uq9j6g"
            style="margin-right: 8px;"
          >
            <span
              class="ant-checkbox ant-wave-target css-dev-only-do-not-override-1uq9j6g"
            >
              <input
                class="ant-checkbox-input"
                type="checkbox"
              />
              <span
                class="ant-checkbox-inner"
              />
            </span>
          </label>
          <span
            style="flex: 1;"
          >
            <span>
              Duty 1
            </span>
          </span>
          <button
            class="ant-btn css-dev-only-do-not-override-1uq9j6g ant-btn-text ant-btn-icon-only"
            type="button"
          >
            <span
              class="ant-btn-icon"
            >
              <span
                aria-label="delete"
                class="anticon anticon-delete"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="delete"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </li>
      <li
        class="is-completed"
      >
        <div
          class="item"
        >
          <label
            class="ant-checkbox-wrapper ant-checkbox-wrapper-checked css-dev-only-do-not-override-1uq9j6g"
            style="margin-right: 8px;"
          >
            <span
              class="ant-checkbox ant-wave-target css-dev-only-do-not-override-1uq9j6g ant-checkbox-checked"
            >
              <input
                checked=""
                class="ant-checkbox-input"
                type="checkbox"
              />
              <span
                class="ant-checkbox-inner"
              />
            </span>
          </label>
          <span
            style="flex: 1;"
          >
            <span>
              Duty 2
            </span>
          </span>
          <button
            class="ant-btn css-dev-only-do-not-override-1uq9j6g ant-btn-text ant-btn-icon-only"
            type="button"
          >
            <span
              class="ant-btn-icon"
            >
              <span
                aria-label="delete"
                class="anticon anticon-delete"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="delete"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </li>
    </ul>
    <div
      class="add-item"
    >
      <span
        aria-label="plus"
        class="anticon anticon-plus"
        role="img"
        style="margin-right: 6px;"
      >
        <svg
          aria-hidden="true"
          data-icon="plus"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
          />
          <path
            d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8z"
          />
        </svg>
      </span>
      <span>
        Add a new item
      </span>
    </div>
  </div>
</div>
`;

exports[`Todos Component displays error message if fetching duties fails 1`] = `
<div>
  <div
    role="alert"
  >
    Failed to load duties
  </div>
</div>
`;

exports[`Todos Component displays loading spinner while fetching duties 1`] = `
<div>
  <div
    role="alert"
  >
    <div
      aria-busy="true"
      aria-live="polite"
      class="ant-spin ant-spin-spinning css-dev-only-do-not-override-1uq9j6g"
    >
      <span
        class="ant-spin-dot-holder"
      >
        <span
          class="ant-spin-dot ant-spin-dot-spin"
        >
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`Todos Component renders AddItem component 1`] = `
<div>
  <div
    id="todos"
  >
    <h3>
      NEX TODO
    </h3>
    <ul
      class="list"
    >
      <li
        class=""
      >
        <div
          class="item"
        >
          <label
            class="ant-checkbox-wrapper css-dev-only-do-not-override-1uq9j6g"
            style="margin-right: 8px;"
          >
            <span
              class="ant-checkbox ant-wave-target css-dev-only-do-not-override-1uq9j6g"
            >
              <input
                class="ant-checkbox-input"
                type="checkbox"
              />
              <span
                class="ant-checkbox-inner"
              />
            </span>
          </label>
          <span
            style="flex: 1;"
          >
            <span>
              Duty 1
            </span>
          </span>
          <button
            class="ant-btn css-dev-only-do-not-override-1uq9j6g ant-btn-text ant-btn-icon-only"
            type="button"
          >
            <span
              class="ant-btn-icon"
            >
              <span
                aria-label="delete"
                class="anticon anticon-delete"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="delete"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </li>
      <li
        class="is-completed"
      >
        <div
          class="item"
        >
          <label
            class="ant-checkbox-wrapper ant-checkbox-wrapper-checked css-dev-only-do-not-override-1uq9j6g"
            style="margin-right: 8px;"
          >
            <span
              class="ant-checkbox ant-wave-target css-dev-only-do-not-override-1uq9j6g ant-checkbox-checked"
            >
              <input
                checked=""
                class="ant-checkbox-input"
                type="checkbox"
              />
              <span
                class="ant-checkbox-inner"
              />
            </span>
          </label>
          <span
            style="flex: 1;"
          >
            <span>
              Duty 2
            </span>
          </span>
          <button
            class="ant-btn css-dev-only-do-not-override-1uq9j6g ant-btn-text ant-btn-icon-only"
            type="button"
          >
            <span
              class="ant-btn-icon"
            >
              <span
                aria-label="delete"
                class="anticon anticon-delete"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="delete"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </li>
    </ul>
    <div
      class="add-item"
    >
      <span
        aria-label="plus"
        class="anticon anticon-plus"
        role="img"
        style="margin-right: 6px;"
      >
        <svg
          aria-hidden="true"
          data-icon="plus"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
          />
          <path
            d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8z"
          />
        </svg>
      </span>
      <span>
        Add a new item
      </span>
    </div>
  </div>
</div>
`;

exports[`Todos Component renders Todos component 1`] = `
<div>
  <div
    id="todos"
  >
    <h3>
      NEX TODO
    </h3>
    <ul
      class="list"
    >
      <li
        class=""
      >
        <div
          class="item"
        >
          <label
            class="ant-checkbox-wrapper css-dev-only-do-not-override-1uq9j6g"
            style="margin-right: 8px;"
          >
            <span
              class="ant-checkbox ant-wave-target css-dev-only-do-not-override-1uq9j6g"
            >
              <input
                class="ant-checkbox-input"
                type="checkbox"
              />
              <span
                class="ant-checkbox-inner"
              />
            </span>
          </label>
          <span
            style="flex: 1;"
          >
            <span>
              Duty 1
            </span>
          </span>
          <button
            class="ant-btn css-dev-only-do-not-override-1uq9j6g ant-btn-text ant-btn-icon-only"
            type="button"
          >
            <span
              class="ant-btn-icon"
            >
              <span
                aria-label="delete"
                class="anticon anticon-delete"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="delete"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </li>
      <li
        class="is-completed"
      >
        <div
          class="item"
        >
          <label
            class="ant-checkbox-wrapper ant-checkbox-wrapper-checked css-dev-only-do-not-override-1uq9j6g"
            style="margin-right: 8px;"
          >
            <span
              class="ant-checkbox ant-wave-target css-dev-only-do-not-override-1uq9j6g ant-checkbox-checked"
            >
              <input
                checked=""
                class="ant-checkbox-input"
                type="checkbox"
              />
              <span
                class="ant-checkbox-inner"
              />
            </span>
          </label>
          <span
            style="flex: 1;"
          >
            <span>
              Duty 2
            </span>
          </span>
          <button
            class="ant-btn css-dev-only-do-not-override-1uq9j6g ant-btn-text ant-btn-icon-only"
            type="button"
          >
            <span
              class="ant-btn-icon"
            >
              <span
                aria-label="delete"
                class="anticon anticon-delete"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="delete"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </li>
    </ul>
    <div
      class="add-item"
    >
      <span
        aria-label="plus"
        class="anticon anticon-plus"
        role="img"
        style="margin-right: 6px;"
      >
        <svg
          aria-hidden="true"
          data-icon="plus"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
          />
          <path
            d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8z"
          />
        </svg>
      </span>
      <span>
        Add a new item
      </span>
    </div>
  </div>
</div>
`;
